File: 20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        d = {'(': 0, ')': 0,
             '[': 1, ']': 1,
             '{': 2, '}': 2}
        for i in s:
            # if we visit the left Parentheses push into stack.
            if i == "(" or i == "[" or i == "{":
                stack.append(i)
            else:
                # we have case that only contain right Parentheses.
                # so it's mean nothing in stack of the case.
                if stack:
                    # pop the last one
                    x = stack.pop()
                    # check if they could combine into a vaild pair.
                    if d[i] == d[x]:
                        continue
                    else:
                        return False
                # nothing in stack so return false.
                else:
                    return False
        # check if any remains in stack.
        # stack is empty then vaild.
        if not stack:
            return True
        # stack is not empty then invaild.
        else:
            return False

File: 20. Valid Parentheses.c


int push(char* stack,char c, int top){
    stack[top] = c;
    return top;
}

char pop(char* stack, int top){
    char data = stack[top];
    return data;
}

bool isEmpty(int top){
    return top == -1 ? true : false;
}

bool isRightOrder(char left, char right){
    
    if (left == '(' && right== ')'){
        return true;
    }else if (left=='[' && right==']'){
        return true;
    }else if (left=='{' && right=='}'){
        return true;
    }
    return false;
}

bool isValid(char * s){
    int top = -1;
    char stack[10001];
    int len = strlen(s);
    for (int i=0;i<len;i++){
        if (s[i]=='(' || s[i]=='[' || s[i]=='{'){
            top = push(stack, s[i],++top);
        }else{
            if (!isEmpty(top)){
                char left = pop(stack, top--);
                if (!isRightOrder(left, s[i])){
                return false;
            }
            }else{
                return false;
            }
        }
    }
    return isEmpty(top);
}
